### PR TITLE
Different file format mount

### LINKED ISSUE #142: Different file format mount
**Describe the solution you'd like**
To mount the configured secrets in different formats, like application.properties for my Java application, JSON/YAML, etc

### PR DESCRIPTION
**Pull request by aristosvo**
**What this PR does / why we need it**:
It generates the secrets in different file formats

**Which issue(s) this PR fixes**:
Fixes #142

### CODE REVIEW

**Comment by aristosvo** [comment]
e2e tests are passing, still needs to add specific tests for the current available file formats:
- JavaProperties
- JSON
- Yaml

File: test/bats/tests/fileformat/nginx-pod-secrets-store-inline-volume-java.yaml [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+  volumes:
+    - name: secrets-store-inline
+      csi:
+        driver: secrets-store.csi.k8s.io
+        readOnly: true
+        volumeAttributes:
+          providerName: "azure"
+          usePodIdentity: "false"         # [OPTIONAL] if not provided, will default to "false"
+          keyvaultName: "$KEYVAULT_NAME" # the name of the KeyVault
+          fileFormatting: "JavaProperties"
```
Comment (aramase) [inline test/bats/tests/fileformat/nginx-pod-secrets-store-inline-volume-java.yaml:22]: Should the `fileFormatting` be defined in the objects array? It's possible the file format is different for different objects in the same `SecretProviderClass`?

Also, the volume attributes are defined in `SecretProviderClass` [example](https://github.com/Azure/secrets-store-csi-driver-provider-azure/blob/master/examples/v1alpha1_secretproviderclass.yaml) and referenced in pods  [here](https://github.com/Azure/secrets-store-csi-driver-provider-azure/blob/master/examples/nginx-pod-secrets-store-inline-volume-secretproviderclass-podid.yaml#L20-L21). The option to define the provider specific parameters in pod spec was deprecated a while back and will be removed in the next release of the driver

Comment (aristosvo) [reply test/bats/tests/fileformat/nginx-pod-secrets-store-inline-volume-java.yaml:22]: Hi @aramase ! Thanks for your comments, as I am definitely willing to improve the implementation.

I think the implementation should not be in the provider specific part, you are right. I will try to improve that part, but it might be harder as providers then all should (try to) support it. What is the right place to discuss these kind of design questions, is there a Slack channel and/or meeting I can join?

With regard to the formatting for the different objects, in my mind it would be confusing to have multiple formatting parameters, especially because the new fileformatting options are combining the objects. It can be done, but it feels a bit like overengineering the feature at this moment in time. It is not that hard to add an extra SecretProviderClass if two formattings are required. WDYT?

Comment (aramase) [reply test/bats/tests/fileformat/nginx-pod-secrets-store-inline-volume-java.yaml:22]: The `spec.parameters` in `SecretProviderClass` is specific to each provider - https://github.com/Azure/secrets-store-csi-driver-provider-azure/blob/master/examples/v1alpha1_secretproviderclass.yaml#L7-L23. `spec.parameters` is a map, so it's extensible for each provider. 

> What is the right place to discuss these kind of design questions, is there a Slack channel and/or meeting I can join?

For driver, we have a slack channel and community meetings - https://github.com/kubernetes-sigs/secrets-store-csi-driver#want-to-help. It would be great if you want to join next week and discuss any changes or questions you have.

> It can be done, but it feels a bit like overengineering the feature at this moment in time. It is not that hard to add an extra SecretProviderClass if two formattings are required. WDYT?

Makes sense!

**Comment by aramase** [comment]
@aristosvo Thank you for opening this PR. Are you still working on this?

**Comment by aristosvo** [comment]
> @aristosvo Thank you for opening this PR. Are you still working on this?

Hi @aramase ! Yes, although I almost forgot about it due to holidays and moving to a different country. I was waiting on #45 to make it easier to initialise my local test setup and add some tests.

**Comment by howlowck** [comment]
Hi @aramase and @aristosvo, we are working on a project that would greatly benefit from this feature.
Is there anything we can help with that can get this PR merged?

Thanks!

**Comment by aramase** [comment]
@howlowck Thanks for the ping. The PR is waiting on #45 for providing scripts for creating objects and testing locally. @bdlb77 has already worked on #45 and there is only one last missing piece. I'll try to take a look at this week and spike on the remaining changes. 

@aristosvo In the meantime could you rebase the PR? We can have the PR in a ready state and run the tests on the CI.

**Comment by howlowck** [comment]
That's great to hear! Thanks @aramase!

**Comment by bdlb77** [comment]
@howlowck Yup! I'm working on one last piece for running the tests correctly in a local environment, so that should be done in the next couple of days.

**Comment by aristosvo** [comment]
@aramase @howlowck I working on it again, but due to the refactoring upstream it took a bit more time. Will work on tests from now on.

**Comment by aramase** [comment]
`Merge branch 'master' into outputfile-formatting` to retrigger broken pipeline jobs.

**Comment by simoncaron** [comment]
@aristosvo Any updates on this? We could really use this feature :)

**Comment by aristosvo** [comment]
@aramase Can your provide me some directions with regard to the unit testing? The e2e tests are a nice improvement and already covering my code more or less.

Do you expect me to break it down in two parts to make unit testing possible or is the coverage okay for now?

I can for instance break it into two parts:
- One part to get the key vault objects from the key vault, store them in an object (which I can cover with unit tests)
- One part writing the key vault objects to file (which are usually not covered with a lot of unit tests, as I/O is involved, but I can do it anyway.)

Another solution might be to cover the current situation in a new unit test which checks written files.

**Comment by aramase** [comment]
> I can for instance break it into:
>  - One part to get the key vault objects from the key vault, store them in an object (which I can cover with unit tests)
>  - One part writing the key vault objects to file (which are usually not covered with a lot of unit tests, as I/O is involved, but I can do it anyway.)

@aristosvo I think this will be a good way to do it. By breaking we can cover it extensively with unit tests. 

Also for the e2e tests, I think we should just have one file with multiple with the `json` and `yaml` tests. WDYT?

**Comment by aristosvo** [comment]
@aramase WDYT about the UT improvements? Is this good enough or should I mock the whole Key Vault Client?

I thought this was good enough for now and covers the scope of this PR.

**Comment by aristosvo** [comment]
@aramase Do you have any feedback? I will rebase this if you think this is still relevant

**Comment by aramase** [comment]
> @aramase Do you have any feedback? I will rebase this if you think this is still relevant

@aristosvo Thanks for the PR. We're making few changes to the driver in the way the files are written to the mount. Essentially the plan is to move the file write logic to the driver instead of individual provider handling it in their own way. So we'll need to make changes to this after that change is implemented.

See [doc](https://docs.google.com/document/d/1y1Oc1_-JGxMh-MZ57iqrVnoCIn6V-W0vJcj3qmq8EWU/edit) for more details.

**Comment by aristosvo** [comment]
Yeah, I knew. Should I open it against the provider? I didn't see a clear timeline for that transformation

**Comment by aramase** [comment]
> Yeah, I knew. Should I open it against the driver? I didn't see a clear timeline for that transformation

The changes for driver writing the files in the pod mount will be available in the next release `v0.0.21` and we'll implement the changes in the provider in the next release. After that we should start the conversation for adding this feature to the driver so there is a central way to handle this. Would you be interested in having a small design write up on how we can implement this in the csi driver?

https://github.com/kubernetes-sigs/secrets-store-csi-driver#want-to-help

**Comment by aristosvo** [comment]
@aramase Do you have a certain format in mind?

From the top of my mind I typed a story-like design doc, don't know if that's good enough? I plan to be on the call next week, but we can discuss and refine before if you'd like.

---
# Secret Store CSI Driver Enhancement Proposal ;)

Multiple applications languages are optimized to consume file based configuration. This is quit different that the Kubernetes Secrets setup, which can be consumed as files (1 secret per file) or as an environment variable. To offer these application formatted files (like JSON) in the current situation we'd use an init container or maybe refered to environment variables from those files.

By enabling an option to mount secrets in a preconfigured file format like JSON, YAML or Java Properties file, we can offer users a better experience.

So, what we have to do is combine multiple secrets, define a file location and a file format and create such a formatted file based on the inputted secrets.An advanced version of this functionality would even offer the possibility to merge with an existing file on that same location, but let's start with the basics.

In a [PR against the Azure provider](https://github.com/Azure/secrets-store-csi-driver-provider-azure/pull/143) this functionality was already implemented, but because the functionality of file mounting is moved to the provider is extended and the role of plugins reduced, this functionality should for the future be implemented in the provider instead of implemented in each seperate plugin.

When looking at that implementation, we see the following hierarchy implemented:

```yaml
apiVersion: secrets-store.csi.x-k8s.io/v1alpha1
kind: SecretProviderClass
metadata:
  name: azure-kvname
  namespace: default
spec:
  provider: azure
  #fileFormatting: "JSON" <- Option 1
  parameters:
    usePodIdentity: "false"
    fileFormatting: "JSON" # position for the Azure plugin
    useVMManagedIdentity: "false"
    userAssignedIdentityID: ""
    keyvaultName: "${KEYVAULT_NAME}"
    objects: |
      array:
        - |
          objectName: secret1              
          objectType: secret
          objectVersion: ""
        - |
          objectName: secret2              
          objectType: secret
          objectVersion: ""
        - |
          objectName: secret3             
          objectType: secret
          objectVersion: ""
        - |
          objectName: cert3             
          objectType: certificate
          objectVersion: ""
    tenantId: "${TENANT_ID}"
---

volumes:
  - name: secrets-store-inline
    csi:
      driver: secrets-store.csi.k8s.io
      readOnly: true
      volumeAttributes:
        secretProviderClass: "my-provider"
       # fileFormatting: "JSON" <- Option 2
```

In the implementation for the Azure plugin the file formatting is , in the parameters which were passed to the plugin. This is one level lower than the first proposed level of configuring the file when doing it in the provider.
But for the provider that's one level too low, so I propose (if we want to put it in the SecretProviderClass) to put it at the spec level. When you want to mount multiple files, I'd propose to those users to use multiple SecretProviderClasses.
From an function point of view it may be better to add it to the volume section, to make it possible to use multiple types of volume from one SecretProviderClass

TL;DR, the functionality can be summarized by:
> Combining and formatting secrets in a file based on 1 variable

**Comment by aramase** [comment]
@aristosvo Thanks for the write up. Could you add that to a google doc and reference it in the meeting agenda? Here are some example design docs: https://github.com/kubernetes-sigs/secrets-store-csi-driver/tree/master/docs/design.

**Comment by TBurda** [comment]
Any updates on this?
